lib/mongo/server/connection_pool.rb

```diff
@@ -513,40 +530,102 @@ def close_idle_sockets
       end
 
       # Create and add connections to the pool until the
-      # pool size is at least min_size.
+      # pool size is at least min_size. Terminates if two connections
+      # encounter errors (in a row) while authenticating.
       #
       # Used by the pool populator background thread.
       #
+      # @return [ true | false ] True if populate ran to completion, false if it
+      # terminated because of errors encountered while connecting connections.
+      #
       # @api private
       def populate
-        return if closed?
+        raise_if_closed!
 
         catch(:done) do
+          retried = false
+
           loop do
+            connection = nil
             @lock.synchronize do
               if !closed? && unsynchronized_size < min_size
-                @available_connections << create_connection
+                connection = create_connection
+                @pending_connections << connection
               else
                 throw(:done)
               end
             end
+
+            begin
+              connect_connection(connection)
+            rescue Exception => e
+              @lock.synchronize do
+                @pending_connections.delete(connection)
+              end
+
+              if retried
+                return false
+              end
+              retried = true
+              next
+            end
+
+            retried = false
+            @lock.synchronize do
+              @available_connections << connection
+              @pending_connections.delete(connection)
+
+              # wake up one thread waiting for connections, since one was created
+              @available_semaphore.signal
+            end
+          end
+        end
+
+        true
+      end
+
+      # Stop the background populator thread and clean up any connections created
+      # by the thread which have not been connected yet.
+      #
+      # @option [ true | false ] wait Wait for background thread to exit before
+      # terminating.
+      def stop_populator(wait = false)
+        @populator.stop!(wait)
+
+        @lock.synchronize do
+          # If stop_populator is called while populate is running, there may be
+          # connections waiting to be connected, connections which have not yet
+          # been moved to available_connections, or connections moved to available_connections
+          # but not deleted from pending_connections. These should be cleaned up.
+          until @pending_connections.empty?
+            connection = @pending_connections.take(1).first
+            connection.disconnect!
+            @pending_connections.delete(connection)
```

Comment:
It could not be creating new connections while this loop is executing, since this loop is locked and so is connection creation (within the populate method). However, it could be trying to connect the connection or waiting for the lock to be released.